server/plugins/FeatureDetection.cpp

```diff
@@ -32,396 +32,544 @@
+        if (dsamp_slope != 0.f) {
+            int nextnsamps = startnsamps + (int)(dsamp_slope * (i+1));
+            int dsamp = nextnsamps - prevnsamps;    // window size delta
+
+            if (dsamp != 0) {
+                float sumChange = 0.;
+
+                if (dsamp > 0) {                    // window grows
+                    for (int j=0; j<dsamp; ++j) {
+                        tail--;                     // grow window
+                        if (tail < 0)
+                            tail += maxsamps;       // wrap
+                        sumChange += data[tail];
+                    }
+                    // add accumulated value overtaken by growing window
+                    // should be outside loop to avoid accumulating error
+                    sum  += sumChange;
+                    sum2 += sumChange;
+                } else {                            // window shrinks: dsamp < 0
+                    int test = abs(dsamp);
+                    for (int j=0; j < test; ++j) {
+                        sumChange += data[tail];
+                        tail++;                     // shrink window
+                        if (tail == maxsamps)
+                            tail = 0;               // wrap
+                    }
+                    // remove sum of values accumulated by shrinking window
+                    // should be outside loop to avoid accumulating error
+                    sum  -= sumChange;
+                    sum2 -= sumChange;
+                }
+
+                prevnsamps += dsamp;
+            }
+        }
+
+        // remove the tail
+        float rmv = data[tail];
+        sum -= rmv;
+
+        // only remove last val from sum2 if sum2 wasn't just reset
+        if (!reset) {
+            sum2 -= rmv;
+            reset = false;
+        }
+
+        // write the new sample in and add it
+        float next= ZXP(in);
+        data[head]= next;
+        sum  += next;
+        sum2 += next;
+
+        ZXP(out) = sum;
+
+        // increment and wrap the head and tail
+        head++;
+        if (head == maxsamps)
+            head = 0;
+        tail++;
+        if (tail == maxsamps)
+            tail = 0;
+        resetcount++;
+
+        // swap the sums once window is full to avoid
+        // floating point error (tip from RunningSum)
+        if (resetcount == nsamps) {
+            sum  = sum2;
+            sum2 = 0.;
+            resetcount = 0;
+            reset = true;
+        }
+    }
+
+    unit->nsamps = nsamps;
+    unit->resetcount= resetcount;
+    unit->head  = head;
+    unit->msum  = sum;
+    unit->msum2 = sum2;
+    unit->tail  = tail;
+    unit->reset = reset;
 }
 
 void initFeatureDetectors(InterfaceTable *it)
 {
-	DefineDtorUnit(PV_JensenAndersen);
-	DefineDtorUnit(PV_HainsworthFoote);
-	DefineDtorUnit(RunningSum);
+    DefineDtorUnit(PV_JensenAndersen);
+    DefineDtorUnit(PV_HainsworthFoote);
+    DefineDtorUnit(RunningSum);
+    DefineDtorUnit(RunningSum2);
```

Comment:
spaces or tabs here

Reply:
@brianlheim I'll try this again from the ground up at some point, but to make sure I have the rules correct: 
* .schelp files should use the tab character (not spaces), set to a 4-space tab.
* .sc class files should do the same
* .cpp UGen source files should use spaces only (not tabs), with a tab distance set to 4 spaces.
Further, I shouldn't touch up any of the rest of the file, even if it doesn't conform to these rules.

I'll try to address each line you've referenced, but tbh I am having trouble tracking which line you mean by "this" or "these" lines when there are multiple lines. So if I have the rules above correct, that will be the guide.

Sorry this is such a time synch, this info on the SC site's contributor's guide would be a useful addition.

Reply:
@snappizz I'll look into how to do unit tests and resubmit, but I'm unfortunately out of time for the next few days :-/

Reply:
sorry I missed this before... Honestly, at this point don't worry too much about the C++ indentation rules. What you've written is correct.

We have a WIP code style document on the Wiki which AFAIK isn't referenced in Contributing.md, but it's reaching a level where we should think about including it: https://github.com/supercollider/supercollider/wiki/%5BWIP%5D-Code-style-guidelines

Thoughts?

Reply:
Great to see this! I hadn't seen this before. This will definitely come in handy, and would definitely be worth linking to from Contributing.md.

Reply:
You'll see a pull request for this from @dyfer shortly...